packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardImage.kt

```diff
@@ -0,0 +1,171 @@
+import android.net.Uri
+import android.os.Build
+import android.provider.MediaStore
+import android.util.Base64
+import androidx.core.os.bundleOf
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runInterruptible
+import java.io.BufferedOutputStream
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.lang.StringBuilder
+import java.util.*
+
+// region Structs and interfaces
+data class ImageResult(val base64Image: String, val width: Int, val height: Int) {
+  fun toBundle() = bundleOf(
+    "data" to base64Image,
+    "size" to bundleOf(
+      "width" to width,
+      "height" to height
+    ),
+  )
+}
+// endregion
+
+// region Module functions
+
+/**
+ * Gets the [imageUri] and returns the [ImageResult] object containing base64 encoded image
+ * and its metadata
+ *
+ * @param context
+ * @param imageUri `content://` uri of the image to be resolved
+ *
+ * @throws IOException
+ * @throws SecurityException when app has no permission to access the content uri
+ */
+internal suspend fun imageFromContentUri(
+  context: Context,
+  imageUri: Uri,
+  options: GetImageOptions
+): ImageResult {
+  // 1. Retrieve bitmap from URI
+  val bitmap = bitmapFromContentUriAsync(context, imageUri)
+
+  // 2. Compress it to target format
+  val format = options.imageFormat
+  val quality = (options.jpegQuality * 100).toInt()
+  val outputStream = ByteArrayOutputStream().also {
+    bitmap.compress(format.compressFormat, quality, it)
+  }
+
+  // 3. Convert to base64
+  val byteArray = outputStream.toByteArray()
+  val encodedString = Base64.encodeToString(byteArray, Base64.DEFAULT)
+  val builder = StringBuilder("data:${format.mimeType};base64,").append(encodedString)
+
+  return ImageResult(
+    base64Image = builder.toString(),
+    width = bitmap.width,
+    height = bitmap.height
+  )
+}
+
+/**
+ * Gets base64-encoded image data and prepares it to be accessible by other applications
+ * when pasting from clipboard
+ *
+ * Saves the image to the [clipboardCacheDir] directory, then uses [ClipboardFileProvider]
+ * to create a `content://` URI which then is packed to the [ClipData] object.
+ *
+ * @param context
+ * @param base64Image base64-encoded JPEG image data. Should not be prefixed
+ * @param clipboardCacheDir directory where the copied image is stored, must be accessible by
+ * the [ClipboardFileProvider]
+ * @return clip data ready to be shared by the [android.content.ClipboardManager]
+ */
+internal suspend fun clipDataFromBase64Image(
+  context: Context,
+  base64Image: String,
+  clipboardCacheDir: File,
+): ClipData {
+  // 1. Get bitmap from base64 string
+  val data = maybeStripBase64Prefix(base64Image)
+  val bitmap = bitmapFromBase64String(data)
+
+  // 2. Create file in cache dir, it will be overwritten if already exists
+  val file = File(clipboardCacheDir, "copied_image.jpeg").also {
+    it.ensureExists()
+  }
```

Comment:
I'm wondering if we should keep this filename constant or randomize it in some way.
- Keeping it constant means that the content URI will always be `content://my.expo.app.ClipboardFileProvider/clipboard_files/copied_image.jpeg` so any app can read it at any time, even weeks after it was copied to the clipboard. Users can clear it manually using expo-file-system by clearing `cacheDir/.clipboard`.
- Randomizing it, e.g. `${randomUUID}.jpeg` means it will be different with each copy, but we need to clear the `clipboardCacheDir` regularly e.g. before copying another image or it will get clogged.

Reply:
Does clipboard offer a single slot for any other data? I guess that's true, so I think we can make this path even simpler `content://my.expo.app.ClipboardFileProvider/image.jpeg` right?
I think having a single image is fine 👌 
We can try to think about for how long this image should be sitting there? 🤔 Maybe there're events that trigger clipboard cleanup for other data types? 🤔 If so maybe we can try to subscribe to them and perform cleanup?

Separate question: is `cacheDir/.clipboard` publicly visible? I mean, is anybody able to locate this directory without doing some manual filesystem traversing?

And last question: is it a problem for us if this file would be accessible forever using this `content://` URI?

Reply:
- We can support copying multiple images in the future.
- Simplifying the URL would be possible if modified the whole `PathStrategy` from the provider. Currently it corresponds to the values defined in `packages/expo-clipboard/android/src/main/res/xml/clipboard_provider_paths.xml` (see the file below)
- > Maybe there're events that trigger clipboard cleanup for other data types?
   🤷 But I don't think it's worth it - if the clipboard was cleared, the URI is no longer there. You can treat it like you forgot URL (removed bookmark) to the website you liked. You don't remember its address but the website is still there.
- `cacheDir/.clipboard` isn't visible anywhere - the whole `ContentProvider` logic is a proxy between the `content://` URI and the real file path.
- I don't think it's a big issue that this uri is always accessible. If somebody really cares, the file can be removed with `expo-file-system` easily. The problem is only when clipboard generates many files which are never deleted - they consume more and more disk space

Reply:
Ok, so right now single file is supported and that's fine 👍 
Cleaning scenarios/other approaches for naming these files are out of scope of this PR 👍